### ISSUE TITLE
Horizon not showing Rabbitmq consummed jobs in Dashboard

### ISSUE THREAD

**Issue by vittorio-c** [vyuldashev/laravel-queue-rabbitmq#345]
Hi,

I don't get any metrics/feedbacks in the Horizon dashboard when I successfully consume a message with the `php artisan horizon` command. 

The consumed job does not appear in the "**Dashboard**", nor in the "**Recent jobs**" section, nor in the "**Metrics**" sction. 

The worker dedicated to rabbitmq and the queue name show up in the Dashboard. 

NB : The Dashboard gets filled with data when I consume a job with 'redis' connection, instead of 'rabbitmq'. So it works outside rabbitmq. 

Could someone help me with that please ? Where should I look for ? 

Here are my config : 

- horizon 3.7.2 
- laravel 6.17.1
- laravel-queue-rabbitmq 10.2

`config/horizon.php`

```
<?php

return [
    'domain' => null,
    'path' => 'horizon',
    'use' => 'default',
    'prefix' => env('HORIZON_PREFIX', 'horizon:'),
    'middleware' => ['web'],
    'waits' => [
        'redis:default' => 60,
    ],
    'trim' => [
        'recent' => 60,
        'completed' => 60,
        'recent_failed' => 10080,
        'failed' => 10080,
        'monitored' => 10080,
    ],
    'fast_termination' => false,
    'memory_limit' => 64,
    'environments' => [
        'production' => [
            'supervisor-1' => [
                'connection' => 'redis',
                'queue' => ['default'],
                'balance' => 'simple',
                'processes' => 10,
                'tries' => 1,
            ],
        ],
        'local' => [
            'supervisor-1' => [
                'connection' => 'rabbitmq',
                'queue' => ['messages'],
                'balance' => 'simple',
                'processes' => 3,
                'tries' => 1,
            ],
            'supervisor-2' => [
                'connection' => 'redis',
                'queue' => ['default'],
                'balance' => 'simple',
                'processes' => 3,
                'tries' => 1,
            ],
        ],
    ],
];
```

`config/queue.php`

```
'default' => env('QUEUE_DRIVER', 'sync'),

'connections' => [
// ...
        'rabbitmq' => [
            'driver' => 'rabbitmq',
            'queue' => env('RABBITMQ_QUEUE', 'default'),
            'connection' => PhpAmqpLib\Connection\AMQPLazyConnection::class,
            'hosts' => [
                [
                    'host' => env('RABBITMQ_HOST', '127.0.0.1'),
                    'port' => env('RABBITMQ_PORT', 5672),
                    'user' => env('RABBITMQ_USER', 'guest'),
                    'password' => env('RABBITMQ_PASSWORD', 'guest'),
                    'vhost' => env('RABBITMQ_VHOST', '/'),
                ],
            ],

            'options' => [
                'exchange' => [
                    'name' => env('RABBITMQ_EXCHANGE_NAME'),
                    'declare' => env('RABBITMQ_EXCHANGE_DECLARE', true),
                    'type' => env('RABBITMQ_EXCHANGE_TYPE', 'fanout'),
                    'passive' => env('RABBITMQ_EXCHANGE_PASSIVE', false),
                    'durable' => env('RABBITMQ_EXCHANGE_DURABLE', true),
                    'auto_delete' => env('RABBITMQ_EXCHANGE_AUTODELETE', false),
                    'arguments' => env('RABBITMQ_EXCHANGE_ARGUMENTS'),
                ],

                'queue' => [
                    'declare' => env('RABBITMQ_QUEUE_DECLARE', true),
                    'bind' => env('RABBITMQ_QUEUE_DECLARE_BIND', true),
                    'passive' => env('RABBITMQ_QUEUE_PASSIVE', false),
                    'durable' => env('RABBITMQ_QUEUE_DURABLE', true),
                    'exclusive' => env('RABBITMQ_QUEUE_EXCLUSIVE', false),
                    'auto_delete' => env('RABBITMQ_QUEUE_AUTODELETE', false),
                    'arguments' => env('RABBITMQ_QUEUE_ARGUMENTS'),
                ],

                'sleep_on_error' => env('RABBITMQ_ERROR_SLEEP', 5),

                'ssl_options' => [
                    'cafile' => env('RABBITMQ_SSL_CAFILE', null), // celui-ci
                    'local_cert' => env('RABBITMQ_SSL_LOCALCERT', null),
                    'local_key' => env('RABBITMQ_SSL_LOCALKEY', null),
                    'verify_peer' => env('RABBITMQ_SSL_VERIFY_PEER', true), // ou ici à false (moins secure)
                    'passphrase' => env('RABBITMQ_SSL_PASSPHRASE', null),
                ],
            ],

            'worker' => env('RABBITMQ_WORKER', 'default'),
        ],

]
```

`config/database.php` : 

```
// other connections 
    'redis' => [
        'client' => 'predis',
        'default' => [
            'host' => env('REDIS_HOST', '127.0.0.1'),
            'password' => env('REDIS_PASSWORD', null),
            'port' => env('REDIS_PORT', 6379),
            'database' => 0,
        ],
    ],
```

 `.env` : 

```
QUEUE_DRIVER=redis
RABBITMQ_WORKER=horizon
RABBITMQ_QUEUE=messages
RABBITMQ_HOST=111.222.333.444
RABBITMQ_PORT=5672
RABBITMQ_USER=user1
RABBITMQ_PASSWORD=password1
RABBITMQ_EXCHANGE_NAME=messages
RABBITMQ_EXCHANGE_TYPE=fanout
RABBITMQ_EXCHANGE_DURABLE=true
RABBITMQ_QUEUE_DECLARE=true
RABBITMQ_QUEUE_DECLARE_BIND=true
RABBITMQ_QUEUE_DURABLE=true
RABBITMQ_ERROR_SLEEP=false
```

**Comment by adm-bome** [comment]
Publish your job to the correct connection!
I think you publish your job to the `redis` connection by default and there is no worker configured within horizon, for the queue `messages` for the redis connection/driver.

within the job constructor try to add this:
```php
   $this->onConnection('rabbitmq')->onQueue('messages')
```
or when dispatching the job tell the dispatcher to use the rabbitmq connection and queue

```php
   TestJob::dispatch()->onConnection('rabbitmq')->onQueue('messages')
```

**Comment by vittorio-c** [comment]
Thank you @adm-bome for your response.

What do you mean by "Publish your job to the correct connection" ? 

I have tried to add ` $this->onConnection('rabbitmq')->onQueue('messages')` on the job constructor, but this does not seem to affect Horizon. 

I publish my jobs through Rabbitmq panel, with the "Publish message" option, on the queue "messages", with no headers, no properties, only the following payload : 

```
{
	"job": "App\\Jobs\\TestJob@handle",
	"data": {
		"key1": "Value 1"
	},
	"id": "8HCJIvi8qSCG7jOdDPelkj21p55pJEjb"
}
```

I suppose the 'job' key allows the worker to call the correct `TestJob@handle`. I'm not sure it's the way to go, though.  Should I add more information to this payload ? 

Here is my TestJob.php : 

```php
<?php

namespace App\Jobs;

use Illuminate\Bus\Queueable;
use Illuminate\Contracts\Queue\ShouldQueue;
use Illuminate\Foundation\Bus\Dispatchable;
use Illuminate\Queue\InteractsWithQueue;
use Illuminate\Queue\SerializesModels;
use Illuminate\Support\Facades\Log;

class TestJob implements ShouldQueue
{
    use Dispatchable, InteractsWithQueue, Queueable, SerializesModels;

    public function __construct()
    {
        $this->onConnection('rabbitmq')->onQueue('messages');
    }

    public function handle($job, $payload)
    {
        try {
            var_dump($payload);
            var_dump(get_class($job));
            $job->delete();
        } catch (\Throwable $e) {
            Log::warning($e->getMessage());
            $job->release();
        }
    }
}
```

I get the correct var_dumps when I run `php artisan horizon` on the console, but nothing on the Horizon web interface.

**Comment by vittorio-c** [comment]
I have added : 

```
    public function handle($job, $payload)
    {
        dump($this->connection);
        dump($this->queue);
        // rest of the code
     }
```

In my job's handler.

When I put `$this->onConnection('rabbitmq')->onQueue('messages')` into the constructor, I get the correct values : "rabbitmq" and "messages". 

When I remove this line in the constructor, I get "null" and "null" for both dumps.

**Comment by adm-bome** [comment]
This is because horizon also works with redis.
Essential data is missing for horizon to work with the message. Horizon can work but mostly works correct when the job was also published from laravel.

When you create a controler in laravel and dispatch the job from within laravel a bigger payload is created. 
```php //WelcomeController.php
<?php

namespace App\Http\Controllers;

class WelcomeController extends Controller
{
    /**
     * Show the application dashboard.
     *
     * @return \Illuminate\Http\Response
     */
    public function index()
    {
        TestJob::dispatch()->onConnection('rabbitmq')->onQueue('messages');

        return view('welcome');
    }
}

```
The message should be put into rabbitMQ.

Most of laravel consumers worker work the way laravel creates a payload. 
A typical payload would be:

```json
{
  "uuid": "6616c749-89a1-4d74-acfb-4ec26eb49794",
  "displayName": "App\\Jobs\\TestJob",
  "job": "Illuminate\\Queue\\CallQueuedHandler@call",
  "maxTries": null,
  "maxExceptions": null,
  "delay": null,
  "timeout": null,
  "timeoutAt": null,
  "data": {
    "commandName": "App\\Jobs\\TestJob",
    "command": "O:16:\"App\\Jobs\\TestJob\":8:{s:3:\"job\";N;s:10:\"connection\";N;s:5:\"queue\";N;s:15:\"chainConnection\";N;s:10:\"chainQueue\";N;s:5:\"delay\";N;s:10:\"middleware\";a:0:{}s:7:\"chained\";a:0:{}}"
  },
  "id": "8HCJIvi8qSCG7jOdDPelkj21p55pJEjb"
}
```

**Comment by adm-bome** [comment]
> I have added :
> 
> ```
>     public function handle($job, $payload)
>     {
>         dump($this->connection);
>         dump($this->queue);
>         // rest of the code
>      }
> ```
> 
> In my job's handler.
> 
> When I put `$this->onConnection('rabbitmq')->onQueue('messages')` into the constructor, I get the correct values : "rabbitmq" and "messages".
> 
> When I remove this line in the constructor, I get "null" and "null" for both dumps.

This is static information var logged. you dump constructor information. unusefull dumps ;)

**Comment by vittorio-c** [comment]
Thank you @adm-bome for your help and patience.

I have tried to put a complete Json Payload in RabbitMQ, and indeed it works as expected. I needed to make sure that the key `data.command` contains correct information (correct number of letter) : 

```
{
  "uuid": "6616c749-89a1-4d74-acfb-4ec26eb49792",
  "displayName":"App\\Jobs\\TestJob",
  "job":"Illuminate\\Queue\\CallQueuedHandler@call",
  "maxTries":null,
  "delay":null,
  "timeout":null,
  "timeoutAt":null,
  "data": {
    "commandName":"App\\Jobs\\TestJob",
    "command":"O:16:\"App\\Jobs\\TestJob\":8:{s:6:\"\u0000*\u0000job\";N;s:10:\"connection\";s:8:\"rabbitmq\";s:5:\"queue\";s:8:\"messages\";s:15:\"chainConnection\";N;s:10:\"chainQueue\";N;s:5:\"delay\";N;s:10:\"middleware\";a:0:{}s:7:\"chained\";a:0:{}}"
  },
  "id":"151",
  "type":"job",
  "tags":[],
  "pushedAt":""
}
```

I got this payload by dumping a job generated by Laravel itself, and modifying it accordingly. 

So, it is indeed recognized by Horizon. I can see the job in Horizon dashbord.

However, this is not an ideal solution. It is too hard to maintain, because we're asking the Application A that publish to RabbitMQ to modelize the payload that way, for Application B (the consumer) in Laravel. Application A don't use Laravel. 

So instead, what we will do, is to call a second Job B from Job A. Job A will be called with bare minimum information "from" RabbitMQ, and will generate a properly Laravel formatted Job B. Job B will have the domain logic, and be watched by Horizon. 

That way we avoid to ask Application A to format payload for Laravel.

Thank you for your help again ! 

I close the issue, because you gave me all the information needed to solve it. 

Best
